runtime/util/hshelp.c

```diff
@@ -1701,6 +1704,75 @@ fixJNIRefs(J9VMThread * currentThread, J9HashTable * classPairs, BOOLEAN fastHCR
 }
 
 
+typedef struct J9ThreadHashPair {
+	J9VMThread *currentThread;
+	void *userData;
+} J9ThreadHashPair;
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+void
+fixMemberNames(J9VMThread * currentThread, J9HashTable * classHashTable)
+{
+	PORT_ACCESS_FROM_JAVAVM(currentThread->javaVM);
+
+	J9ThreadHashPair data;
+	data.currentThread = currentThread;
+	data.userData = classHashTable;
+
+	/* iterate over all objects fixing up vmtarget refs if object is a MemberName */
+	currentThread->javaVM->memoryManagerFunctions->j9mm_iterate_all_objects(currentThread->javaVM, PORTLIB, 0, fixMemberNamesObjectIteratorCallback, &data);
+}
+
+static jvmtiIterationControl
+fixMemberNamesObjectIteratorCallback(J9JavaVM *vm, J9MM_IterateObjectDescriptor *objectDesc, void *userData)
+{
+	J9ThreadHashPair *data = userData;
+	J9VMThread *currentThread = data->currentThread;
+	J9HashTable *classHashTable = data->userData;
+	j9object_t object = objectDesc->object;
+	J9Class *clazz = J9OBJECT_CLAZZ_VM(vm, object);
+
+	if (NULL != classHashTable) {
+		if (clazz == J9VMJAVALANGINVOKEMEMBERNAME(vm) && 0 != J9OBJECT_U64_LOAD(currentThread, object, vm->vmindexOffset)) {
```

Comment:
Make this nested ifs so you can fetch the vmIndex only once.

Also, the macro to use here is `J9VMJAVALANGINVOKEMEMBERNAME_OR_NULL` to avoid attempting to load the class.

Reply:
Wouldn't fetching the vmindex the second time be optimized out by the compiler anyway? Since it's the same expression

Reply:
Maybe or maybe not - we're not careful about const and volatile use.